msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/OidcAuthority.java

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class OidcAuthority extends Authority {
+    //Part of the OpenIdConnect standard, this is appended to the authority to create the endpoint that has OIDC metadata
+    static final String WELL_KNOWN_OPENID_CONFIGURATION = ".well-known/openid-configuration";
+    private static final String AUTHORITY_FORMAT = "https://%s/%s/";
+
+    OidcAuthority(URL authorityUrl) throws MalformedURLException {
+        super(transformAuthority(authorityUrl), AuthorityType.OIDC);
+
+        this.authority = String.format(AUTHORITY_FORMAT, host, tenant);
+    }
+
+    private static URL transformAuthority(URL originalAuthority) throws MalformedURLException {
+        String transformedAuthority = originalAuthority.toString();
+        transformedAuthority += WELL_KNOWN_OPENID_CONFIGURATION;
+
+        return new URL(transformedAuthority);
+    }
+
+    void setAuthorityProperties(AadInstanceDiscoveryResponse instanceDiscoveryResponse) {
+        this.authorizationEndpoint = instanceDiscoveryResponse.authorizationEndpoint();
```

Comment:
I think you are conflating instance discovery and oidc discovery. These are different http calls, with different responses etc. Instance Discovery is an AAD invention. OIDC discovery is based on the OIDC protocol. 

For AAD, B2C, ADFS and CIAM, MSAL4J chose not skip OIDC discovery and hardcode the endpoints.
For CIAM CUD / OIDC authority, this is no longer possible. OIDC discoevry needs to happen.

Reply:
I recommend you create more abstraction to model this complexity. For example you can model both through interfaces InstanceDiscovery and OidcDiscovery. MSAL would use both to compute urls. And provide 2 implementations for each: 

- NoInstanceDiscovery - which applies to everything but AAD. 
- AadInstanceDiscovery - which applies to AAD

- HardcodedOidcDIscovery - which applies to everything but OIDC
- OidcDiscovery - which applies to OIDC

Thoughts?

Reply:
In the latest commit I moved the OIDC discovery logic into some new classes, and restored the AAD instance discovery classes to what they were before these changes. Now the instance discovery classes only do their behavior when the authority type is AAD, and and the OIDC discovery classes only do their behavior when the authority type is OIDC.

However, what do you mean by "HardcodedOidcDiscovery - which applies to everything but OIDC"? Do all other authority types need to make a call to some hardcoded version of that ".well-known/openid-configuration" endpoint?